src/plugins/kibana_utils/common/persistable_state/types.ts

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { SavedObjectReference } from '../../../../core/types';
+
+/**
+ * Serializable state is something is a POJO JavaScript object that can be
+ * serialized to a JSON string.
+ */
+export type SerializableState = {
+  [key: string]: Serializable;
+};
+export type SerializableValue = string | number | boolean | null | undefined | SerializableState;
+export type Serializable = SerializableValue | SerializableValue[];
+
+/**
+ * Versioned state is a POJO JavaScript object that can be serialized to JSON,
+ * and which also contains the version information. The version is stored in
+ * semver format and corresponds to the Kibana release version when the object
+ * was created. The version can be used to apply migrations to the object.
+ *
+ * For example:
+ *
+ * ```ts
+ * const obj: VersionedState<{ dashboardId: string }> = [{ dashboardId: '123' }, '7.14.0'];
+ * ```
+ */
+export type VersionedState<S extends SerializableState = SerializableState> = [
+  state: S,
```

Comment:
do we want to add this nesting ?
should we also add id ?

Reply:
I would rather have nesting, than force some properties on user's object. Basically I think composition (nesting) is better than "inheritance".

Currently it is:

```ts
type VersionedState<S> = [state: S, versions: string];
```

Instead it could be also:

```ts
interface VersionedState<S> {
  versions: string;
  state: S;
}
```

I thing both choices above of "nesting" are better than forcing shape of user's object:

```ts
type VersionedState<S> = S & { version: string };
// or
type VersionedState<S> = S & { __version: string };
```

---

With regards to version: yes, it sounds like it would be useful too. We also discussed saved object references before. Maybe Persistable State service could have a "wrapper" that captures all three: (1) version; (2) id; (3) references.

```ts
interface MigratableState<S> {
  state: S;
  id?: string;
  version?: string;
  references?: SavedObjectReference[];
}
```

We could introduce `MigratableState` first in limited use, for example, only `migrateToLatest` would use it at first. Then create tickets to upgrade all Persistable State service methods to use it eventually.